KAFKA-14516: [1/N] Static Member leave, join, re-join request using ConsumerGroupHeartbeats

vamossagar12 opened the pull request:
This is the first PR which aims at adding static membership to the new consumer group protocol. It aims at a static member joining, re-joining and leaving a consumer group using the new ConsumerGroupHeartbeat API in group coordinator. Note that the rejoining case is within the session timeout and kicking out a static member if no static member joins within session timeout is not part of this PR.

vamossagar12 commented:
@dajac , do you think it is ok to have this small PR focussed only on static member departure? I plan to create such small PRs for other cases when a static member joins etc. Let me know if you prefer this or a single PR with all changes. Thanks!

vamossagar12 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -929,12 +929,22 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+        if (member.instanceId() != null) {
+            log.info("Member {} with instance id {} is a static member and will not be fenced from the group", memberId, member.instanceId());
+        } else {
+            log.info("[GroupId {}] Member {} left the consumer group.", groupId, memberId);
```
I have taken the liberty and updated the log line to use an argument based loggers instead of the string concatenation based pattern that existed before.

vamossagar12 commented:
Test failures seem unrelated.

dajac reviewed (commented):
@vamossagar12 Thanks for the PR. I had a quick look at it and I have a few high level comments:
* I don't mind doing a few small PRs at all.
* The title of the PR is a bit confusing because the member does not send a LeaveGroup request.
* I think that we need to keep the -2 epoch. I think that it could be useful in the future if we want to extend the Consumer to have the ability to chose whether it wants to leave forever or temporarily.
* Ignoring the "leave request" is not enough. If you look at the other side, we need to know whether the member has released the instance id or not. We could perhaps capture this by updating the member epoch to -2 or something like that.

Commit:
Using member epoch -2 for static member leave request

vamossagar12 commented:
Thanks for the review David! Responses inline

> I don't mind doing a few small PRs at all.
Ack. Thanks for the confirmation.

> The title of the PR is a bit confusing because the member does not send a LeaveGroup request.

Sorry about the confusion. I have corrected it.

> I think that we need to keep the -2 epoch. I think that it could be useful in the future if we want to extend the Consumer to have the ability to chose whether it wants to leave forever or temporarily.

Got it. I have reverted to using member epoch -2 for static member leave request. 

> Ignoring the "leave request" is not enough. If you look at the other side, we need to know whether the member has released the instance id or not. We could perhaps capture this by updating the member epoch to -2 or something like that.

I have updated the member epoch to -2 if the departing worker is a static member. I am not sure if that will be enough though.

vamossagar12 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -593,6 +609,9 @@ private void throwIfConsumerGroupHeartbeatRequestIsInvalid(
             }
+        } else if (request.memberEpoch() == -2) {
+            throwIfEmptyString(request.memberId(), "MemberId can't be empty.");
+            throwIfNull(request.instanceId(), "InstanceId can't be empty for Static Member.");
```
This is not necessarily needed but added to avoid situations when a non static member tries to send a member epoch -2.

Commit:
Correcting error when instance id is null for static member

kirktrue commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -593,6 +609,9 @@ private void throwIfConsumerGroupHeartbeatRequestIsInvalid(
                 throw new InvalidRequestException("SubscribedTopicNames must be set in first request.");
             }
+        } else if (request.memberEpoch() == -2) {
+            throwIfEmptyString(request.memberId(), "MemberId can't be empty.");
```
This is for the case where the static member is leaving temporarily, right?

Would it be possible to add more detail to these error messages to aid in troubleshooting/debugging on the client when this condition is hit?

vamossagar12 replied:
> This is for the case where the static member is leaving temporarily, right

Yes, that's correct. I have added some debugging information (like groupId etc). Let me know if that makes sense.

vamossagar12 replied:
Also, `MemberId can't be empty.` string is used in other places as well (when member epoch is > 0  or equal to -1. Should we look to change those as well?

kirktrue replied:
It's just my preference, so if there's precedent for how you have it, I wouldn't hold up this PR in an effort to change the other places. Thanks!

vamossagar12 replied:
Makes sense. I think I updated that one comment. But yeah the rest of the loggers, I won't be touching them, as that would be noisy to review.

kirktrue commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -918,23 +937,32 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+        // The departing member is a static one. We don't need to fence this member because it is
+        // expected to come back within session timeout
+        if (memberEpoch == -2) {
+            log.info("Member {} with instance id {} is a static member and will not be fenced from the group", memberId, member.instanceId());
```
Nice! Can we add the Group ID to the 'static member' log message?

vamossagar12 replied:
Done.

kirktrue reviewed (changes requested):
Thanks for the PR, @vamossagar12! Just a couple of minor changes, mostly around error and log messages.

Commit:
Addressing review comments

vamossagar12 commented:
Thanks for the review @kirktrue ! I have addressed the changes.

vamossagar12 commented:
I realised that a logic similar to https://github.com/apache/kafka/blob/trunk/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/generic/GenericGroup.java#L774-L780 is missing in this PR. I think that's important. To get this working though, I would also need to add the logic to add a different map for staticMembers and also support adding a static member to the group. In that case, I might need to enhance this PR to support both Leave request and join request.

Commit:
Adding logic for static member join

Commit:
Adding static member join or rejoin mechanism with consumer group heartbeat

Commit:
Adding More tests for static member join, re-join and leave requests using ConsumerGroupHeartbeat

Commit:
Resolving merge conflicts

Commit:
Rebasing with latest trunk

vamossagar12 commented:
> I realised that a logic similar to https://github.com/apache/kafka/blob/trunk/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/generic/GenericGroup.java#L774-L780 is missing in this PR. I think that's important. To get this working though, I would also need to add the logic to add a different map for staticMembers and also support adding a static member to the group. In that case, I might need to enhance this PR to support both Leave request and join request.

I updated this PR to handle this case and to also handle (re)-join and leaving the group using ConsumerGroupHeartbeat api.

dajac commented:
Thanks @vamossagar12! I will get to this PR soon.

vamossagar12 commented:
> Thanks @vamossagar12! I will get to this PR soon.

Thank you. I am still coming to terms with the new group coordinator so there might be rough edges in this PR (upfront apologies for that )

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -564,6 +564,22 @@ private void throwIfNotNull(
+     */
+    private void throwIfNull(
+            Object value,
+            String error
```
nit: Indentation should be 4 spaces.

vamossagar12 replied:
done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -107,6 +107,11 @@ public static class DeadlineAndEpoch {
+    /**
+     * The static group members.
+     */
+    private final Map<String, String> staticMembers;
```
I suppose that this must be a `TimelineHashMap`.

vamossagar12 replied:
done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -285,6 +302,64 @@ public ConsumerGroupMember getOrMaybeCreateMember(
+    public ConsumerGroupMember getOrMaybeCreateStaticMember(
+            String memberId,
+            String instanceId,
+            boolean createIfNotExists
```
nit: Indentation should be four spaces.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -285,6 +302,64 @@ public ConsumerGroupMember getOrMaybeCreateMember(
+            // No existing member found against this instance id. Creating new.
+            if (existingMemberId == null) {
+                member = getOrMaybeCreateMember(memberId, true);
+                staticMembers.put(instanceId, memberId);
```
The state should not be updated like this. All the updates are handled in the `replay()` methods.

vamossagar12 replied:
Makes sense, I have removed this direct update of states and moved it to `replay()`

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -285,6 +302,64 @@ public ConsumerGroupMember getOrMaybeCreateMember(
+            // The member joined with a non-zero epoch but we haven't registered this static member
+            // This could be an unknown member for the coordinator.
+            if (existingMemberId == null) {
+                throw Errors.UNKNOWN_MEMBER_ID.exception();
```
Could we add custom message to all the exceptions raise in this method?

vamossagar12 replied:
This method is no longer being used. I added custom messages in the other method which is now being called.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -285,6 +302,64 @@ public ConsumerGroupMember getOrMaybeCreateMember(
+                int currentMemberEpoch = existingMember.memberEpoch();
+                // A new member with a used instance id joined but the previous member using the same instance id
+                // hasn't requested leaving the group.
+                if (currentMemberEpoch != -2 && !existingMemberId.equals(memberId)) {
```
I wonder if we really need the second part of the condition here. What was your thinking about it?

vamossagar12 replied:
Actually this is no longer required. Have removed it.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -285,6 +302,64 @@ public ConsumerGroupMember getOrMaybeCreateMember(
+                }
+                // A new static member is trying to take the place of a departed static member. We will
+                // provide the assignments of the old member to the new one.
+                member = new ConsumerGroupMember.Builder(memberId, existingMember)
+                        .setMemberEpoch(existingMember.targetMemberEpoch())
+                        .setPreviousMemberEpoch(0)
+                        .setTargetMemberEpoch(existingMember.targetMemberEpoch())
+                        .build();
```
This is a bit weird because you pass `existingMember` to the builder and then you still have to override other fields. Would it be better to do `new ConsumerGroupMember.Builder(existingMember)` and then override the fields? I think that we only need to set the new member id.

nit: The indentation should be four spaces.

vamossagar12 replied:
YEs, that didn't quite make sense. It's not needed anymore. Fixed indent as well.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -285,6 +302,64 @@ public ConsumerGroupMember getOrMaybeCreateMember(
+                        .setPreviousMemberEpoch(0)
+                        .setTargetMemberEpoch(existingMember.targetMemberEpoch())
+                        .build();
+                updateMember(member);
+                staticMembers.put(instanceId, memberId);
```
As said previously, the state should not be updated here but in replay.

vamossagar12 replied:
Ack.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -750,7 +770,9 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
         if (memberId.isEmpty()) memberId = Uuid.randomUuid().toString();
-        final ConsumerGroupMember member = group.getOrMaybeCreateMember(memberId, createIfNotExists);
+        final ConsumerGroupMember member = instanceId == null ?
+                group.getOrMaybeCreateMember(memberId, createIfNotExists) :
+                group.getOrMaybeCreateStaticMember(memberId, instanceId, createIfNotExists);
```
nit: indentation.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -750,7 +770,9 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+                group.getOrMaybeCreateStaticMember(memberId, instanceId, createIfNotExists);
         throwIfMemberEpochIsInvalid(member, memberEpoch, ownedTopicPartitions);
 
         if (memberEpoch == 0) {
```
I wonder if we could log something here as well when a static member is replaced.

dajac replied:
I just thought about something else. When a static member is replaced, we need to write records to erase the state of the previous member.

vamossagar12 replied:
> I just thought about something else. When a static member is replaced, we need to write records to erase the state of the previous member.

YEs that was a miss. I have added relevant tombstone records for the replaced static member and also cancelled it's timers.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -790,6 +812,12 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
                     groupId, memberId, updatedMember.subscribedTopicRegex());
                 bumpGroupEpoch = true;
             }
+        } else {
+            // A new static member replaces an older one with the same instance id, assignments etc.
+            // We will create a new member subscription record for this new member.
+            if (instanceId != null) {
+                records.add(newMemberSubscriptionRecord(groupId, updatedMember));
+            }
```
This does not seem correct because we will write a record whenever the member is not updated and we have an instance id. I think that it would be better to capture the fact that we have a new static member in the condition at L801.

vamossagar12 replied:
This bit of code has changed now.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -908,23 +936,40 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+        List<Record> records = new ArrayList<>();
+        // The departing member is a static one. We don't need to fence this member because it is
+        // expected to come back within session timeout
+        if (memberEpoch == -2) {
```
I would rather prefer to have a separate method for the static leave group.

vamossagar12 replied:
Added.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -908,23 +936,40 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+                    groupId, memberId, member.instanceId());
+            // We will write a member epoch of -2 for this departing static member.
+            ConsumerGroupMember leavingStaticMember = new ConsumerGroupMember.Builder(member)
+                    .setMemberEpoch(-2)
+                    .build();
```
nit: indentation.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -1084,10 +1129,14 @@ public CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGro
         throwIfConsumerGroupHeartbeatRequestIsInvalid(request);
 
-        if (request.memberEpoch() == LEAVE_GROUP_MEMBER_EPOCH) {
+        if (request.memberEpoch() == LEAVE_GROUP_MEMBER_EPOCH || request.memberEpoch() == -2) {
```
nit: Should we introduce a constant for -2 as well?

vamossagar12 replied:
done

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -1127,7 +1176,9 @@ public void replay(
 
         if (value != null) {
-            ConsumerGroupMember oldMember = consumerGroup.getOrMaybeCreateMember(memberId, true);
+            ConsumerGroupMember oldMember = value.instanceId() != null ?
+                    consumerGroup.getOrMaybeCreateStaticMember(memberId, value.instanceId(), true) :
+                    consumerGroup.getOrMaybeCreateMember(memberId, true);
```
We don't need to use getOrMaybeCreateStaticMember here as we only want to look up the member by its id.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -1127,7 +1176,9 @@ public void replay(
+            ConsumerGroupMember oldMember = value.instanceId() != null ?
+                    consumerGroup.getOrMaybeCreateStaticMember(memberId, value.instanceId(), true) :
+                    consumerGroup.getOrMaybeCreateMember(memberId, true);
             consumerGroup.updateMember(new ConsumerGroupMember.Builder(oldMember)
```
In there, we need to update the static id mapping in updateMember and removeMember, I think.

vamossagar12 replied:
Done

dajac reviewed (commented):
@vamossagar12 Thanks for PR! I just made a first pass on it and I left some comments. I don't have a clear picture of the end state yet so I need a bit more time to think about it. However, my first comments are still valid, I think.

Commit:
Addressing review comments

Commit:
Adding log line when static member joins or rejoins with 0 epcoh

Commit:
Removing unwanted method

vamossagar12 commented:
Thanks @dajac , I have addressed the comments.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -259,6 +265,17 @@ public void setTargetAssignmentEpoch(int targetAssignmentEpoch) {
+     * instance id.
+     *
+     * @param groupInstanceId the group instance id.
+     * @return the static member if it exists.
```
nit: We tend to use a single line for getters. eg. `@return The member id corresponding to the given instance id or null if it does not exist`.

vamossagar12 replied:
Done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -285,6 +302,11 @@ public ConsumerGroupMember getOrMaybeCreateMember(
         return member;
     }
 
+    public ConsumerGroupMember getStaticMember(String instanceId) {
```
nit: We don't prefix getters with `get`. Let's add javadoc as well.

vamossagar12 replied:
Done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -299,6 +321,9 @@ public void updateMember(ConsumerGroupMember newMember) {
         maybeUpdateServerAssignors(oldMember, newMember);
         maybeUpdatePartitionEpoch(oldMember, newMember);
         maybeUpdateGroupState();
+        if (newMember.instanceId() != null) {
+            staticMembers.put(newMember.instanceId(), newMember.memberId());
+        }
```
nit: Would it make sense to have a method like the others? I would also do this before calling `maybeUpdateGroupState`.

vamossagar12 replied:
Done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -312,6 +337,9 @@ public void removeMember(String memberId) {
         maybeUpdateServerAssignors(oldMember, null);
         maybeRemovePartitionEpoch(oldMember);
         maybeUpdateGroupState();
+        if (oldMember.instanceId() != null) {
+            staticMembers.remove(oldMember.instanceId());
+        }
```
ditto.

vamossagar12 replied:
Done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -107,6 +107,11 @@ public static class DeadlineAndEpoch {
      */
```
Let's add unit tests for the new or changed methods to the corresponding file.

vamossagar12 replied:
Done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -983,27 +1052,125 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+        try {
+            TargetAssignmentBuilder.TargetAssignmentResult assignmentResult =
+                new TargetAssignmentBuilder(groupId, groupEpoch, assignors.get(preferredServerAssignor))
+                    .withMembers(group.members())
```
I think that the old member will be in `members` so the computed target assignment is incorrect. We need to remove it with `removeMember` and we also need to set the target assignment of the new member from the old one.

vamossagar12 replied:
I missed the `removeMembers` part. Thanks for pointing it out. Regarding 

> we also need to set the target assignment of the new member from the old one.

I assumed, this [block of code](https://github.com/apache/kafka/blob/trunk/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java#L262-L288)  should take care of it. Was my assumption wrong?

dajac replied:
Well, at the moment, a new target assignment will be computed for the new static member and that block of code will indeed create the record for it. What I meant is that the new static member should actually reuse the target assignment of the previous member (vs computing a new one). Then we only need to recompute the assignment if there is a group epoch bump.

vamossagar12 replied:
That makes sense and thanks for the explanation. I have now changed the code to remove the existing static member and add the new static member. Rest of the state would remain as is.

vamossagar12 replied:
Also, regarding 

> Then we only need to recompute the assignment if there is a group epoch bump.

Given that now we will have a group epoch bump whenever a static member re-joins with a different subscription, should this be mentioned in the KIP? As we noticed, this is a deviation from how the static member rejoining with a different subscription case as of today. WDYT?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -983,27 +1052,125 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+        ConsumerGroupMember existingStaticMember,
+        ConsumerGroupMember updatedMember
+    ) {
+        // Write tombstones for the departed static member. Follow the same order as the one followed when fencing
+        // a member
+        records.add(newCurrentAssignmentTombstoneRecord(group.groupId(), existingStaticMember.memberId()));
+        records.add(newTargetAssignmentTombstoneRecord(group.groupId(), existingStaticMember.memberId()));
+        records.add(newMemberSubscriptionTombstoneRecord(group.groupId(), existingStaticMember.memberId()));
```
We have similar code somewhere else. Could we add a method for this and reuse it?

vamossagar12 replied:
Done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -983,27 +1052,125 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+        records.add(newCurrentAssignmentTombstoneRecord(group.groupId(), existingStaticMember.memberId()));
+        records.add(newTargetAssignmentTombstoneRecord(group.groupId(), existingStaticMember.memberId()));
+        records.add(newMemberSubscriptionTombstoneRecord(group.groupId(), existingStaticMember.memberId()));
+        // Cancel all the timers of the departed static member.
+        cancelConsumerGroupSessionTimeout(group.groupId(), existingStaticMember.memberId());
+        cancelConsumerGroupRevocationTimeout(group.groupId(), existingStaticMember.memberId());
```
Same for this one. It would be great to have a method.

vamossagar12 replied:
Done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -983,27 +1052,125 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+        cancelConsumerGroupSessionTimeout(group.groupId(), existingStaticMember.memberId());
+        cancelConsumerGroupRevocationTimeout(group.groupId(), existingStaticMember.memberId());
+        // Write a record corresponding to the new member
+        records.add(newMemberSubscriptionRecord(groupId, updatedMember));
+        TargetAssignmentBuilder.TargetAssignmentResult assignmentResult = computeTargetAssignment(group, group.groupEpoch(), member, updatedMember);
+        records.addAll(assignmentResult.records());
+        records.add(newCurrentAssignmentRecord(groupId, updatedMember));
```
I am not sold on this. Is it too difficult to reuse the main logic? There are a few issues with this approach. For instance, the member's assignment is not reconciled like we do in the main logic. Another one is that the subscription metadata must be updated as well if the subscriptions have changed.

vamossagar12 replied:
The main reason for extracting this out was that while using the main logic, there always always a group epoch bump even when a new static member replaces an older one. When I debugged it further, it seems to be because of this logic [here](https://github.com/apache/kafka/blob/trunk/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java#L885-L889)
. More specifically, the issue at hand is that `subscriptionMetadata` has partition racks info while the currently stored metadata doesn't have it. This is with regards to the test `testStaticMemberGetsBackAssignmentUponRejoin`. I  wasn't totally sure if this is an issue with the test itself but since this led to a group epoch bump, I thought we shouldn't do it. 

Actually when I think about it now, maybe it makes sense to have a group epoch bump in this case as well. While it might go against no rebalance during static member rejoin but the reason for rejoin is a change in subscription metadata and not a static member re-join. The latter seemed harder to replicate via tests though because it always bumped up the group epoch due to the above mentioned issue. Please let me know your thoughts.

dajac replied:
Hum... I am not sure to fully follow. The subscription metadata should not be different if the subscriptions and the metadata image have not changed. Does the new static member has the same subs as the previous one in your case?

vamossagar12 replied:
I got around the issue by explicitly setting the rack info in the subscription metadata like [here](https://github.com/apache/kafka/pull/14432/files#diff-1396a400e5c9f5ccaa44cb20194066eaa919a7e02c558db02005a4a2b67a93b9R2117-R2122). I guess so far this wasn't apparent because all the tests expected a group epoch bump happening. In this case, we didn't want a group epoch bump and hence I could notice the discrepancy.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -747,6 +798,11 @@ private void throwIfMemberEpochIsInvalid(
         List<ConsumerGroupHeartbeatRequestData.TopicPartitions> ownedTopicPartitions
     ) {
         if (receivedMemberEpoch > member.memberEpoch()) {
+            // If a static member rejoins, it's previous epoch would be -2. In such a
+            // case, we don't need to fence the member.
+            if (member.memberEpoch() == LEAVE_GROUP_STATIC_MEMBER_EPOCH && receivedMemberEpoch == 0) {
+                return;
+            }
```
I don't fully get this one. Could you please elaborate?

vamossagar12 replied:
This is mainly needed for the static member rejoin case. Let's say a static member with instance id `id` departed. When it departs, we would write a member epoch value of -2 against it. Now, if a new static member joins with the same instance id `id` and a member epoch value of 0, then without this condition, the rejoin would always fail with `FencedMemberEpochException`. This condition was added to avoid the same.

dajac replied:
I see. Would it make sense to put this condition first in the method, before `if (receivedMemberEpoch > member.memberEpoch())`? I got confused by the fact that it is within the if branch.

vamossagar12 replied:
Yeah, that makes sense as well. I placed it inside the if condition because this condition at hand shows up only when received epoch > current member epoch. But it should be ok to have it outside as you said. I have made the change.

dajac reviewed (commented):
@vamossagar12 Thanks for the update. I left a few more comments.

vamossagar12 commented:
Thanks for the review @dajac , Most of my changes are on my local, but before I push those, wanted your thoughts on [this](https://github.com/apache/kafka/pull/14432/files#r1363301832) comment. I can proceed with the rest of the changes post that.

Commit:
Addressing some review comments

dajac commented:
@vamossagar12 Thanks. I just replied to your questions.

Commit:
Addressing further review tests + adding unit tests in ConsumerGroupTest

vamossagar12 commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,715 @@ public void testLeavingMemberBumpsGroupEpoch() {
+            .setRebalanceTimeoutMs(5000)
+            .setSubscribedTopicNames(Arrays.asList("foo", "bar"))
+            .setServerAssignorName("range")
+            .setPartitionsPendingAssignment(mkAssignment(
```
@dajac , I realised that even in the static member re-joining case, while the group epoch doesn't bump, the partitions would be in pending assignment state. I believe, eventually the member would get it's assignments. In that case, this state seems correct to me. WDYT?

vamossagar12 commented:
Thanks @dajac , I have addressed all the comments. Have added a couple of more questions. Thanks for all the pointers, the changes look cleaner than the previous iteration (to me atleast).

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -829,20 +885,49 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
         // Get or create the member.
         if (memberId.isEmpty()) memberId = Uuid.randomUuid().toString();
-        final ConsumerGroupMember member = group.getOrMaybeCreateMember(memberId, createIfNotExists);
+        ConsumerGroupMember member;
+        ConsumerGroupMember existingStaticMember = null;
+        if (instanceId == null) {
+            member = group.getOrMaybeCreateMember(memberId, createIfNotExists);
+        } else {
+            existingStaticMember = group.staticMember(instanceId);
+            throwIfStaticMemberValidationFails(groupId, instanceId, existingStaticMember, memberEpoch, memberId);
+            member = group.getOrMaybeCreateMember(memberId, createIfNotExists);
+        }
```
I wonder if we could simplify it even more. For instance, would it be possible to have something like the following:

```
ConsumerGroupMember member;
ConsumerGroupMember updatedMember;

if (instanceId == null) {
  member = group.getOrMaybeCreateMember(memberId, createIfNotExists);
  throwIfMemberEpochIsInvalid(member, memberEpoch, ownedTopicPartitions);
  log.info(...);
  updatedMember = new ConsumerGroupMember.Builder(member)
     ....
} else {
  // the new logic.
  // member is the current static member.
  // updatedMember is the updated current member or the new one.
}
```

vamossagar12 replied:
I have updated the logic in line with your suggestion.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -869,6 +954,10 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
                 bumpGroupEpoch = true;
             }
+        } else {
+            if (staticMemberReplaced) {
+                records.add(newMemberSubscriptionRecord(groupId, updatedMember));
+            }
```
If we rely on `member` and `updatedMember` then we don't need this because `!updatedMember.equals(member)` will catch the new member.

vamossagar12 replied:
Yes, that makes sense.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -898,31 +987,44 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+                // We just need to remove the older member and add the newer one. The new member can
+                // reuse the target assignment of the older member.
+                if (staticMemberReplaced && groupEpoch == targetAssignmentEpoch) {
+                    targetAssignment = group.targetAssignment(existingStaticMember.memberId());
+                    assignmentResult = assignmentResultBuilder
+                        .removeMember(existingStaticMember.memberId())
+                        .addOrUpdateMember(memberId, updatedMember)
+                        .build();
+                    records.addAll(assignmentResult.records());
```
I don't fully understand how this would work because the members and the target assignment are not set.

vamossagar12 replied:
I have added members and target assignment.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -898,31 +987,44 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+                    assignmentResult = assignmentResultBuilder
                         .withMembers(group.members())
                         .withSubscriptionMetadata(subscriptionMetadata)
                         .withTargetAssignment(group.targetAssignment())
                         .addOrUpdateMember(memberId, updatedMember)
```
When the `TargetAssignmentBuilder` builds the spec for the assignor, it must use the target assignment of the previous static member for the new static member. How do we ensure this? We may have to update the `TargetAssignmentBuilder` to understand that a static member is replaced.

vamossagar12 replied:
I think I got it now. I added some state tracking to the `TargetAssignmentBuilder` so that it doesn't compute assignments for a replacing static member.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -983,27 +1085,81 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
         return new CoordinatorResult<>(records, response);
```
Don't we need to also force the step 3.? If we don't do it, we don't write the current assignment record for the new member and we don't reconcile him.

vamossagar12 replied:
Yes, I am now building the replacing static member with the same set of assignments (target/pending). This forces it directly to have the current assignment record for the new member.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -983,27 +1085,81 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
         return new CoordinatorResult<>(records, response);
     }
 
+    private void replaceStaticMemberInConsumerGroup(
```
nit: removeMemberAndCancelTimers? The logic is not tight to static members. I would also directly pass the groupId and the memberId as this is all it needs.

vamossagar12 replied:
done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -983,27 +1085,81 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
-        List<Record> records = consumerGroupFenceMember(group, member);
+        List<Record> records = new ArrayList<>();
+        if (memberEpoch == LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
+            throwIfStaticMemberValidationFails(groupId, instanceId, member, memberEpoch, memberId);
```
This is actually executed twice. Once here and once in `consumerGroupStaticMemberGroupLeave`. I also wonder if we need to full validation here. I suppose that ensuring that the member id is correct would be enough, no?

vamossagar12 replied:
Removed.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -983,27 +1085,81 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
-
-        List<Record> records = consumerGroupFenceMember(group, member);
+        List<Record> records = new ArrayList<>();
+        if (memberEpoch == LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
```
nit: We check this twice. Once here and once earlier to lookup the member. Could we combine them?

vamossagar12 replied:
done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -983,27 +1085,81 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+        List<Record> records = new ArrayList<>();
+        if (memberEpoch == LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
+            throwIfStaticMemberValidationFails(groupId, instanceId, member, memberEpoch, memberId);
+            log.info("[GroupId {}] Member {} with instance id {} is a static member and will not be fenced from the group",
+                    group.groupId(), member.memberId(), member.instanceId());
```
nit: `"[GroupId {}] Static member {} with member id {} left the consumer group."`? I would also use a similar logging structure for the other log messages.

vamossagar12 replied:
done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -983,27 +1085,81 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+     * instance id decided to leave the group and would be back within session
+     * timeout.
+     *
+     * @param group       The group.
+     * @param existingStaticMember      The member.
```
nit: Let's align the description of the params.

vamossagar12 replied:
done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -983,27 +1085,81 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+     */
+    private List<Record> consumerGroupStaticMemberGroupLeave(
+        ConsumerGroup group,
+        ConsumerGroupMember existingStaticMember,
```
nit: `member`?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -983,27 +1085,81 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+            records.addAll(consumerGroupStaticMemberGroupLeave(group, member, memberId));
+        } else {
+            log.info("[GroupId {}] Member {} left the consumer group.", groupId, memberId);
+            records.addAll(consumerGroupFenceMember(group, member));
```
nit: The `addAll` does not seem necessary here. Could we avoid it?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:
```diff
@@ -287,6 +305,18 @@ public ConsumerGroupMember getOrMaybeCreateMember(
+    /**
+     * Gets a static member.
+     *
+     * @param instanceId        The group instance id.
```
nit: We can reduce the space between the name and the desc.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -671,6 +688,10 @@ private void throwIfConsumerGroupHeartbeatRequestIsInvalid(
                 throw new InvalidRequestException("SubscribedTopicNames must be set in first request.");
             }
+        } else if (request.memberEpoch() == LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
+            throwIfEmptyString(request.memberId(), "MemberId can't be empty. GroupId: " + request.groupId());
```
nit: Let's use `"MemberId can't be empty."` to be consistent with the previous errors.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -671,6 +688,10 @@ private void throwIfConsumerGroupHeartbeatRequestIsInvalid(
             }
+        } else if (request.memberEpoch() == LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
+            throwIfEmptyString(request.memberId(), "MemberId can't be empty. GroupId: " + request.groupId());
+            throwIfNull(request.instanceId(), "InstanceId can't be null for Static Member. GroupId: "
```
I think that the instance id cannot be null and cannot be empty as well. Then let's use `InstanceId can't be null or empty.`

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -731,6 +752,36 @@ private void throwIfConsumerGroupIsFull(
+                log.info("Request memberId={} for static member with groupInstanceId={} " +
+                        "is fenced by existing memberId={}",
+                    memberId, instanceId, existingStaticMember.memberId());
+                throw Errors.FENCED_INSTANCE_ID.exception();
```
Could we add something to the exception?

vamossagar12 replied:
Added.

dajac reviewed (commented):
@vamossagar12 Thanks for the last update. I just made a pass on it and left a few more comments. I think that we are almost there!

Commit:
Addressing more review comments

vamossagar12 commented:
Thanks @dajac , I have addressed the review comments. Please let me know how the changes are looking now.

Commit:
Adding unit tests for TargetAssignmentBuilde static member replace case

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -671,6 +688,10 @@ private void throwIfConsumerGroupHeartbeatRequestIsInvalid(
             }
+        } else if (request.memberEpoch() == LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
+            throwIfEmptyString(request.memberId(), "MemberId can't be empty.");
+            throwIfNull(request.instanceId(), "InstanceId can't be null or empty for Static Member. GroupId: "
```
I would just use `InstanceId can't be null.` here. I think that we should also verify that the instance id is not empty.

vamossagar12 replied:
sure. `instance id should not be empty` check is already happening [here](https://github.com/apache/kafka/blob/trunk/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java#L658)

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -829,21 +902,50 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
-            log.info("[GroupId {}] Member {} joins the consumer group.", groupId, memberId);
+        ConsumerGroupMember member;
+        ConsumerGroupMember.Builder updatedMemberBuilder;
+        ConsumerGroupMember updatedMember;
```
nit: It seems that we could declare this one only when we use it at L948.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -829,21 +902,50 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+            }
+            updatedMemberBuilder = new ConsumerGroupMember.Builder(member);
+        } else {
+            member = group.staticMember(instanceId);
```
I still find this logic quite complex to follow. I wonder if we could be a little more explicit. I think that the complexity comes from `throwIfStaticMemberValidationFails` which hide quite a lot of the logic.

I wonder if something as follow would be better. I am not sure... What do you think?

```
ConsumerGroupMember existingMember = group.staticMember(instanceId);
if (memberEpoch == 0) {
   // A new static member joins or the existing static member rejoins.
   if (existingMember == null) {
      // New static member.
      member = group.getOrMaybeCreateMember(memberId, true);
      updatedMemberBuilder = new ConsumerGroupMember.Builder(member);
   } else {
      // Static member rejoins with a different instance id so it should replace
      // the previous instance iff the previous instance has -2.
      if (existingMember.memberEpoch() != LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
         // The new member can't join.
         throw Errors.UNRELEASED_INSTANCE_ID.exception(...);
      } else {
         // Replace the current member.
        staticMemberReplaced = true;
        member = existingMember;
        updatedMemberBuilder = new ConsumerGroupMember.Builder(group.getOrMaybeCreateMember(memberId, true));
        removeMemberAndCancelTimers(records, group.groupId(), member.memberId());
      }
} else {
   // Check member id or throw FENCED_INSTANCE_ID
   // Check epoch with throwIfMemberEpochIsInvalid
}
```

Note that I just wrote this without testing it so the code is likely not 100% correct :).

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -829,21 +902,50 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+            throwIfMemberEpochIsInvalid(member, memberEpoch, ownedTopicPartitions);
+            if (createIfNotExists) {
+                log.info("[GroupId {}] Member {}, Instance-Id {} joins the consumer group.", groupId, memberId, instanceId);
+            }
```
I would say `[GroupId {}] Static member {} with instance id {} joins the consumer group.`

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -829,21 +902,50 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+                removeMemberAndCancelTimers(records, group.groupId(), member.memberId());
+                // The replacing member gets the same set of assignments as the departed member.
+                updatedMemberBuilder = new ConsumerGroupMember.Builder(memberId)
+                    .setAssignedPartitions(member.assignedPartitions())
+                    .setPartitionsPendingAssignment(member.partitionsPendingAssignment())
+                    .setPartitionsPendingRevocation(member.partitionsPendingRevocation());
```
I was thinking about this a little more and I actually wonder if we need this after all. If we don't see it, I think that the reconciliation will kick in and compute the current assignment. What do you think?

vamossagar12 replied:
If I don't set `setAssignedPartitions`, then the replacing static member doesn't get it's assignments back. The pending assignments bit, I added just to ensure all assignments from previous member are assigned to the new member. The reason that happens is that it lands up [here](https://github.com/apache/kafka/blob/6b26c0428a5cf1f64e458e5a33df3bbfb0e33c6b/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/CurrentAssignmentBuilder.java#L175) and if there are no partitions assigned, it gets back empty assignments.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -898,31 +1000,45 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+                // A new static member is replacing an older one with the same subscriptions.
+                // We just need to remove the older member and add the newer one. The new member can
+                // reuse the target assignment of the older member.
+                if (staticMemberReplaced && groupEpoch == targetAssignmentEpoch) {
```
Is the condition really correct? It seems to me that we must remove the previous member whenever staticMemberReplaced is true.

How about the following?

```
TargetAssignmentBuilder assignmentResultBuilder =
    new TargetAssignmentBuilder(groupId, groupEpoch, assignors.get(preferredServerAssignor))
         .withMembers(group.members())
         .withSubscriptionMetadata(subscriptionMetadata)
         .withTargetAssignment(group.targetAssignment());

if (staticMemberReplaced) {
    assignmentResultBuilder
         .removeMember(member.memberId())
         .addOrUpdateMember(memberId, updatedMember);
} else {
    assignmentResultBuilder
         .addOrUpdateMember(memberId, updatedMember);
}

TargetAssignmentBuilder.TargetAssignmentResult assignmentResult = 
    assignmentResultBuilder.build();
```

vamossagar12 replied:
Actually, `groupEpoch == targetAssignmentEpoch` is not needed. I was just trying to ensure that the group epoch and target member epoch are the same which is what will happen when static member is replaced. So, in a way it's redundant. I will remove it.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -898,31 +1000,45 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
                 targetAssignmentEpoch = groupEpoch;
+                log.info("[GroupId {}] Computed a new target assignment for epoch {}: {}.",
+                    groupId, groupEpoch, assignmentResult.targetAssignment());
+
```
nit: Let's keep this log message where it was.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -983,27 +1099,81 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
-        ConsumerGroupMember member = group.getOrMaybeCreateMember(memberId, false);
-
-        log.info("[GroupId " + groupId + "] Member " + memberId + " left the consumer group.");
+        ConsumerGroupMember member = memberEpoch == LEAVE_GROUP_STATIC_MEMBER_EPOCH ?
```
How about the following?

```
List<Record> records;
if (instanceId == null) {
  ConsumerGroupMember member = group.getOrMaybeCreateMember(memberId, false);
  log.info("[GroupId {}] Member {} left the consumer group.", groupId, memberId);
  records = consumerGroupFenceMember(group, member);
} else {
  // Get static member
  ...
  // Validate the member id matches the expect id or throw fenced member id
  ...
  if (memberEpoch == LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
      log.info("[GroupId {}] Static Member {} with instance id {} temporarily left the consumer group",
           group.groupId(), memberId, instanceId);
      records = consumerGroupStaticMemberGroupLeave(group, instanceId, member, memberId);
  } else {
      log.info("[GroupId {}] Static Member {} with instance id {} left the consumer group",
           group.groupId(), memberId, instanceId);
      records = consumerGroupFenceMember(group, member);
  }
}
```

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java:
```diff
@@ -195,6 +200,9 @@ public TargetAssignmentBuilder addOrUpdateMember(
     ) {
+        if (member != null && member.instanceId() != null) {
+            this.staticMembers.put(memberId, member.instanceId());
+        }
```
Instead of doing this, could we just pass the mapping from the ConsumerGroup?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java:
```diff
@@ -221,22 +230,36 @@ public TargetAssignmentBuilder removeMember(
      */
     public TargetAssignmentResult build() throws PartitionAssignorException {
         Map<String, AssignmentMemberSpec> memberSpecs = new HashMap<>();
+        Map<String, Assignment> assignmentByInstanceId = new HashMap<>();
```
It is a tad annoying that we have to build this mapping here because we usually only need it for one member. Instead of doing this, I wonder if we could lookup the member id from the instance id mapping and then get the target assignment of the member id. The mapping will contain the previous member or nothing.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+                    put(memberId2, new MemberAssignment(mkAssignment(
+                        mkTopicAssignment(fooTopicId, 3, 4, 5),
+                        mkTopicAssignment(barTopicId, 2)
+                    )));
```
Should we remove this as the member is not there anymore?

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+        ConsumerGroupMember member2UpdatedEpoch = new ConsumerGroupMember.Builder(member2)
+            .setMemberEpoch(-2)
+            .build();
+        // The departing static member will have it's epoch set to -2.
```
nit: Let's put this comment before `ConsumerGroupMember member2UpdatedEpoch = ...`.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+
+        assertEquals(result.records(), expectedRecords);
+        // Replay the updated record with -2 epoch so that the group can update itself.
+        context.replay(RecordHelpers.newCurrentAssignmentRecord(groupId, member2UpdatedEpoch));
```
If I recall correctly, we automatically do this in `consumerGroupHeartbeat()`.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+                .setSubscribedTopicNames(Arrays.asList("foo", "bar"))
+                .setTopicPartitions(Collections.emptyList()));
+
+        // member epoch of the response would be set to -2
```
nit: `Member...`.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+        // Replay the updated record with -2 epoch so that the group can update itself.
+        context.replay(RecordHelpers.newCurrentAssignmentRecord(groupId, member2UpdatedEpoch));
+
+        // member 2 rejoins the group with the same instance id
```
nit: `Member...` and `.` at the end.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+                    )));
+                }
+            }
+            ));
```
I suppose that this is not required as it don't be called.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+
+    @Test
+    public void testShouldThrowInvalidRequestExceptionWhenInstanceIdIsNullForStaticMember() {
+
```
nit: Let's remove this empty line.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroupTest.java:
```diff
@@ -99,6 +128,28 @@ public void testRemoveMember() {
+        assertFalse(consumerGroup.hasMember("member"));
+        assertNull(consumerGroup.staticMember("instance"));
+        assertNull(consumerGroup.staticMemberId("instance"));
+
```
nit: We can remove this empty line.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+                ))),
+                rejoinResult.response()
+        );
+
```
Should we also verify that the timers are cancelled correctly?

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+    }
+
+    @Test
+    public void testShouldThrownUnreleasedInstanceIdExceptionWhenMemberJoinsWithInUseInstanceId() {
```
Do we test all possible error cases? Thinking of fenced instance id, unknown member id, fenced member epoch, etc.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+    }
+
+    @Test
+    public void testNoGroupEpochBumpWhenStaticMemberLeaves() {
```
This is true if the member leaves with -2. If it leaves with -1, it should be removed immediately. Should we test this as well?

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+                .setSubscribedTopicNames(Arrays.asList("foo", "bar"))
+                .setTopicPartitions(Collections.emptyList())));
+    }
+
```
I wonder if we could add a few more tests. Thinking about the following ones:
* The leaving static member should disappear if the new one does not rejoin with the session timeout.
* The new data structures should be updated correctly on replay.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+                .addRacks()
+                .build())
+            .withConsumerGroup(new ConsumerGroupBuilder(groupId, 10)
+                .withMember(new ConsumerGroupMember.Builder(memberId1)
```
Should we simplify a bit this test? We only need on member in the group to verify what we want here. We could also use one topic only, etc.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+    }
+
+    @Test
+    public void testShouldThrowFencedInstanceIdExceptionWhenStaticMemberWithDifferentMemberIdLeaves() {
```
We could also simplify this one.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1908,6 +1909,723 @@ public void testLeavingMemberBumpsGroupEpoch() {
+    }
+
+    @Test
+    public void testShouldThrowUnknownMemberIdExceptionWhenUnknownStaticMemberLeaves() {
```
This one as well.

dajac reviewed (commented):
@vamossagar12 Thanks for the update and your patience on this one! We're almost there, I think. I left some more comments for consideration.

Commit:
Addressing review comments

vamossagar12 committed:
Merge branch 'trunk' into KAFKA-14516-static-member-leave

vamossagar12 commented:
Thanks @dajac . I have addressed all the comments. The second part of [this](https://github.com/apache/kafka/pull/14432/files#r1397382111) comment, I am not sure which new data structures being referred here. The test `testStaticMemberGetsBackAssignmentUponRejoin` asserts all the records returned after replay. Is that what you are referring to?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -691,6 +708,10 @@ private void throwIfConsumerGroupHeartbeatRequestIsInvalid(
             }
+        } else if (request.memberEpoch() == LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
+            throwIfEmptyString(request.memberId(), "MemberId can't be empty.");
+            throwIfNull(request.instanceId(), "InstanceId can't be null for Static Member. GroupId: "
```
I would rather use `InstanceId can't be null.` here in order to be consistent with the other error messages.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -751,6 +772,53 @@ private void throwIfConsumerGroupIsFull(
+     * but the existing static member hasn't requested leaving the group.
+
+     */
+    private void throwIfStaticMemberValidationFails(
```
I suppose that we could remove this one now. Could we?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -766,6 +834,11 @@ private void throwIfMemberEpochIsInvalid(
+        // case, we don't need to fence the member.
+        if (member.memberEpoch() == LEAVE_GROUP_STATIC_MEMBER_EPOCH && receivedMemberEpoch == 0) {
+            return;
+        }
```
Is it still necessary with the last implementation?

vamossagar12 replied:
No it is not. Removed it.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -849,21 +922,53 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+                    log.info("[GroupId {}] Static member {} with instance id {} joins the consumer group", groupId, memberId, instanceId);
+                } else {
+                    // Static member rejoins with a different member id so it should replace
+                    // the previous instance iff the previous member had sent a Leave group.
```
nit: `Leave` -> `leave`.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -849,21 +922,53 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+                    staticMemberReplaced = true;
+                    updatedMemberBuilder = new ConsumerGroupMember.Builder(memberId)
+                        .setAssignedPartitions(member.assignedPartitions());
+                    removeMemberAndCancelTimers(records, group.groupId(), member.memberId());
```
Should we also log something in this case?

vamossagar12 replied:
Added a log line

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -849,21 +922,53 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+                }
+            } else {
+                throwIfStaticMemberIsUnknown(member, instanceId);
+                throwIfInstanceIdIsFenced(memberId, instanceId, member);
```
nit: I would put `member` as the first argument to be consistent with the other helpers.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -1004,27 +1118,102 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
         return new CoordinatorResult<>(records, response);
     }
 
+    private void throwIfInstanceIdIsUnreleased(String groupId, String memberId, String instanceId, ConsumerGroupMember member) {
```
* Could we move those helpers next to `throwIfMemberEpochIsInvalid`? Could we also add some javadoc to each of them?
* I wonder if we could also find better names for the params because it is not clear whether `memberId` and `instanceId` are the ones of the existing member or the ones received in the request. We could perhaps use `receivedMemberId`, etc. What do you think? This also applies to the other helpers.

vamossagar12 replied:
Done, moved the methods next to `throwIfMemberEpochIsInvalid`, added Javadocs and updated the argument names.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -1004,27 +1118,102 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+    private void throwIfInstanceIdIsFenced(String memberId, String instanceId, ConsumerGroupMember member) {
+        if (!member.memberId().equals(memberId)) {
+            log.info("Request memberId={} for static member with groupInstanceId={} " +
+                    "is fenced by existing memberId={}",
```
I wonder if we could follow the structure of the other log messages here: `[GroupId {}] Static Member {} with instance id {}....`.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -1004,27 +1118,102 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
 
+    private void throwIfInstanceIdIsUnreleased(String groupId, String memberId, String instanceId, ConsumerGroupMember member) {
+        if (member.memberEpoch() != LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
+            // The new member can't join.
```
Should we also log something here?

vamossagar12 replied:
Done.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -1004,27 +1118,102 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+            throwIfStaticMemberIsUnknown(member, instanceId);
+            throwIfInstanceIdIsFenced(memberId, instanceId, member);
+            if (memberEpoch == LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
+                log.info("[GroupId {}] Static Member {} with instance id {} temporarily left the consumer group",
```
nit: `.` at the end.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -1004,27 +1118,102 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+                    group.groupId(), memberId, instanceId);
+                records.add(consumerGroupStaticMemberGroupLeave(group, member));
+            } else {
+                log.info("[GroupId {}] Static Member {} with instance id {} left the consumer group",
```
nit: `.` at the end.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -1004,27 +1118,102 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+            if (memberEpoch == LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
+                log.info("[GroupId {}] Static Member {} with instance id {} temporarily left the consumer group",
+                    group.groupId(), memberId, instanceId);
+                records.add(consumerGroupStaticMemberGroupLeave(group, member));
```
nit: Could we also use `records = ` here? With this, we could remove `new ArrayList<>()` when `records` is declared, I think.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java:
```diff
@@ -223,20 +238,30 @@ public TargetAssignmentResult build() throws PartitionAssignorException {
-            targetAssignment.getOrDefault(memberId, Assignment.EMPTY),
-            subscriptionMetadata
-        )));
+        members.forEach((memberId, member) -> {
```
nit: Let's revert this change as it is not necessary.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java:
```diff
@@ -223,20 +238,30 @@ public TargetAssignmentResult build() throws PartitionAssignorException {
+                ConsumerGroupMember member = members.get(memberId);
+                Assignment assignment;
+                // A new static member joins and needs to replace an existing departed one.
+                if (member == null && staticMembers.containsKey(updatedMemberOrNull.instanceId())) {
+                    assignment = targetAssignment.getOrDefault(staticMembers.get(updatedMemberOrNull.instanceId()), Assignment.EMPTY);
```
I am not sure to follow this one. My understanding is that we populate `staticMembers` only when `addOrUpdateMember` is called. In the main flow, we basically call this only once with the new or updated member.

Let's imagine that a new static member joins. We will add its static id with its member id to `staticMembers`. Therefore here, we basically get back its member id and end up with no assignment. Did I get this right?

I think that this could work but we would need to pass the `staticMembers` mapping from the `ConsumerGroup` to the builder, like we pass the members. If we have this, we could use it here to find the previous member with the static id if the member is new and has a static id.

vamossagar12 replied:
Yes, your understanding is correct. I see what you are saying about how this won't work when a new static member joins. I have updated the group to expose the current set of static members in the group.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -1004,27 +1118,102 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+
+    private void throwIfStaticMemberIsUnknown(ConsumerGroupMember staticMember, String instanceId) {
+        if (staticMember == null) {
+            throw Errors.UNKNOWN_MEMBER_ID.exception("No existing static member found against instance id: " +  instanceId);
```
nit: `Instance id {} is unknown.`?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -1004,27 +1118,102 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+        if (member.memberEpoch() != LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
+            // The new member can't join.
+            throw Errors.UNRELEASED_INSTANCE_ID.exception("Member " + memberId + " trying to join group "
+                + groupId + " but the instance id " + instanceId + " is already in use by member " + member.memberId());
```
nit: `Static member {} with instance id {} cannot join the group because the instance id is owned by member {}.`?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -1004,27 +1118,102 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+                    "is fenced by existing memberId={}",
+                memberId, instanceId, member.memberId());
+            throw Errors.FENCED_INSTANCE_ID.exception("Member " + memberId + " for static member with groupInstanceId " + instanceId +
+                " is fenced by existing memberId " + member.memberId());
```
nit: `Static member {} with instance id {} was fenced by member {}.`?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -849,21 +922,53 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+                    throwIfInstanceIdIsUnreleased(groupId, memberId, instanceId, member);
+                    // Replace the current member.
+                    staticMemberReplaced = true;
+                    updatedMemberBuilder = new ConsumerGroupMember.Builder(memberId)
+                        .setAssignedPartitions(member.assignedPartitions());
```
To close on this one, it is indeed correct to set the assigned partitions here. Without it, the reconciler checks if the partitions in the target assignment are still owned and they are effectively still owned until the previous member is removed. This only happens when the records are processed.

vamossagar12 replied:
Thank you for the confirmation.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -1004,27 +1118,102 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+    ) {
+        // We will write a member epoch of -2 for this departing static member.
+        ConsumerGroupMember leavingStaticMember = new ConsumerGroupMember.Builder(member)
+            .setMemberEpoch(LEAVE_GROUP_STATIC_MEMBER_EPOCH)
```
I think that we could also `setPartitionsPendingRevocation` to empty because we know that the member has revoked all its partitions when it leaves.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilderTest.java:
```diff
@@ -691,6 +735,91 @@ public void testDeleteMember() {
+
+        context.updateMemberSubscription("member-1", Arrays.asList("foo", "bar", "zar"), Optional.of("member-1"), Optional.empty());
+        context.updateMemberSubscription("member-2", Arrays.asList("foo", "bar", "zar"), Optional.of("member-2"), Optional.empty());
+        context.updateMemberSubscription("member-3", Arrays.asList("foo", "bar", "zar"), Optional.of("member-3"), Optional.empty());
```
This should not be here. I think that you mix in two different things. `addGroupMember` is basically what is used to build what will be passed to `withMembers` and `withTargetAssignment` whereas `updateMemberSubscription` is for `addOrUpdateMember`. Therefore, the test does not reproduce how we use it.

vamossagar12 replied:
I see. Thanks for the explanation, I hadn't understood the usage of these methods correctly. I have removed these unwanted calls to `updateMemberSubscription`

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilderTest.java:
```diff
@@ -691,6 +735,91 @@ public void testDeleteMember() {
+        context.removeMemberSubscription("member-3", "member-3");
+
+        // Another static member joins with the same instance id as the departed one
+        context.addGroupMember("member-3-a", "member-3", Arrays.asList("foo", "bar", "zar"), new HashMap<>());
```
This one is incorrect as well because the newly added member is not added via `withMembers`.

vamossagar12 replied:
Ok.. I am slightly confused by this comment. The new member is being added using `addGroupMember` which internally invokes `withMembers`. I had an unwanted call to `updateMemberSubscription` which I have removed. Probably I am missing something here.

dajac replied:
I think that we should call `updateMemberSubscription` instead of calling `addGroupMember` here because `updateMemberSubscription` is what is used to eventually call `addOrUpdateMember`.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1919,6 +1920,879 @@ public void testLeavingMemberBumpsGroupEpoch() {
+                    )));
+                }
+            }
+            ));
```
nit: Indentation seems off here. I think that it should be 4 spaces earlier.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1919,6 +1920,879 @@ public void testLeavingMemberBumpsGroupEpoch() {
+                        .setPartitions(Arrays.asList(3, 4, 5)),
+                    new ConsumerGroupHeartbeatResponseData.TopicPartitions()
+                        .setTopicId(barTopicId)
+                        .setPartitions(Collections.singletonList(2))
```
nit: Indentation is incorrect.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1919,6 +1920,879 @@ public void testLeavingMemberBumpsGroupEpoch() {
+            .setServerAssignorName("range")
+            .setAssignedPartitions(mkAssignment(
+                    mkTopicAssignment(fooTopicId, 3, 4, 5),
+                    mkTopicAssignment(barTopicId, 2)))
```
nit: Indentation.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1919,6 +1920,879 @@ public void testLeavingMemberBumpsGroupEpoch() {
+                .setGroupId(groupId)
+                .setMemberId(memberId2)
+                .setInstanceId(memberId2)
+                .setMemberEpoch(-2)
```
Let's replace `-2` with the relevant constant. There are other cases in this file.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1919,6 +1920,879 @@ public void testLeavingMemberBumpsGroupEpoch() {
+    }
+
+    @Test
+    public void testNoGroupEpochBumpWhenStaticMemberLeaves() {
```
nit: Temporarily leave?

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1919,6 +1920,879 @@ public void testLeavingMemberBumpsGroupEpoch() {
+    }
+
+    @Test
+    public void testShouldThrowInvalidRequestExceptionWhenInstanceIdIsNullForStaticMember() {
```
We already have `testConsumerHeartbeatRequestValidation` so I wonder if we could just add the new case there. What do you think?

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -3328,6 +4202,87 @@ public void testSessionTimeoutExpiration() {
+
+        // Static member sends a temporary leave group request
+        result = context.consumerGroupHeartbeat(
+                new ConsumerGroupHeartbeatRequestData()
```
nit: Indentation.

dajac reviewed (commented):
@vamossagar12 Thanks for the update! I reviewed the entire PR and it looks pretty good. I think that I have one major questions around the `TargetAssignmentBuilder` changes where I think that it is not quite right. All the other are small comments/nits.

Commit:
Addressing more review comments

Commit:
Correcting log line

vamossagar12 commented:
@dajac , thank you for another round of review. I have handled all review comments.

Commit:
Addressing missed review comments

dajac reviewed (approved):
LGTM, thanks @vamossagar12!
